### PR TITLE
Add seeing is believing layer

### PR DESCRIPTION
**Pull request by brandoncc**

### CODE REVIEW

File: layers/+tools/seeing-is-believing/packages.el [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+      ))
+
+
+(defun seeing-is-believing/init-seeing-is-believing ()
+  (use-package seeing-is-believing
+    :defer t
+    :init
+    (progn
+      (require 'seeing-is-believing)
+      (add-hook 'ruby-mode-hook 'seeing-is-believing))
```
Comment (TheBB) [inline layers/+tools/seeing-is-believing/packages.el]: If you require `seeing-is-believing` in the `:init` section, you are essentially making it not deferred and loaded at startup every time.

**Comment by TheBB** [comment]
Sorry, but this `use-package` call isn't very up to scratch. `:init` runs before loading and `:config` runs after loading, but since you load the package in `:init`, everything ends up running at the same time, causing this package to be loaded on startup. Remove the require and make the keybindings in `:init` instead of `:config`.

**Comment by nixmaniack** [comment]
I believe `seeing-is-believing` belongs to `ruby` layer instead separate layer.

**Comment by brandoncc** [comment]
@TheBB Thank you for the feedback. I have a better understanding of `use-package` now. I had not used it at all before I got this working, and the only way I could make it work was to do it the way I had it. `require` is no longer necessary now that I am using only `init`. I did not understand why it was necessary, but that is precisely because of my `defer`.

@nixmaniack I didn't put it there because it is a tool that not everybody uses, should I still put it there? I should put some sort of condition on it to only init if the executable is present I assume? If so, could you point me in the right direction to learn about that and where exactly the condition should be?

**Comment by nixmaniack** [comment]
@brandoncc Not sure about the best way, others should correct me.

You should look at `ruby-layer`'s `packages.el` and how each package is initialised. `use-package` has [conditional loading](https://github.com/jwiegley/use-package#conditional-loading) which you can make use of with `(executable-find "seeing-is-believing")` for loading only if binary is found.

**Comment by brandoncc** [comment]
Ah, I did read that conditional loading section but I didn't see anything
about an executable (I still don't but I will try it with the code you
provided). I was using the ruby layer as my example when I was building it
actually, I just wasn't looking at the most relevant section I guess.
Thanks for the help.

**Comment by brandoncc** [comment]
@TheBB I believe I have addressed the issues. Please take a look.

**Comment by dcluna** [comment]
@brandoncc :+1: This is cool

File: layers/+lang/ruby/packages.el [hunk 2]
```diff
@@ -126,3 +127,18 @@
...
+    :commands (seeing-is-believing seeing-is-believing-run seeing-is-believing-clear)
+    :if (executable-find "seeing_is_believing")
+    :init
+    (progn
+      (add-hook 'ruby-mode-hook 'seeing-is-believing)
+      (evil-leader/set-key-for-mode 'ruby-mode "msb" 'seeing-is-believing-run)
+      (evil-leader/set-key-for-mode 'enh-ruby-mode "msb" 'seeing-is-believing-run)
+      (evil-leader/set-key-for-mode 'ruby-mode "msc" 'seeing-is-believing-clear)
+      (evil-leader/set-key-for-mode 'enh-ruby-mode "msc" 'seeing-is-believing-clear)))
+  )
```
Comment (dcluna) [inline layers/+lang/ruby/packages.el]: Lisp convention does not allow trailing parentheses like this one. Can you move it to the end of the previous line?

Comment (brandoncc) [inline layers/+lang/ruby/packages.el]: Fixed, but the PR is out of date and can't be merged. I don't use spacemacs anymore so I won't be fixing that unfortunately. Anybody else is welcome to.

**Comment by brandoncc** [comment]
@TheBB I fixed the merge issues on this.

**Comment by scepticulous** [comment]
I am happy somebody is already working on this. Would be great to have this being part of spacemacs,
especially since I believe it is still not an elpa package.

**Comment by Compro-Prasad** [comment]
According to [comment](https://github.com/syl20bnr/spacemacs/pull/3436#discussion_r54960468) by @brandoncc what should be done t this PR @sdwolfz @syl20bnr .

File: layers/+lang/ruby/README.org [hunk 3]
```diff
@@ -164,6 +166,13 @@ When =ruby-test-runner= equals =ruby-test=.
 | ~SPC m t b~ | run test file       |
 | ~SPC m t t~ | run test at pointer |
 
+** seeing-is-believing
+
+| Key Binding     | Description                      |
+|-----------------+----------------------------------|
+| ~<SPC> m s b~   | Run seeing is believing          |
+| ~<SPC> m s c~   | Clear seeing is believing output |
+
```
Comment (syl20bnr) [inline layers/+lang/ruby/README.org]: Need to find other key bindings, these one are reserved for REPL interactions.
Can those two functions be unified together into a toggle ?

Comment (Compro-Prasad) [reply layers/+lang/ruby/README.org]: This ain't changing.

Comment (brandoncc) [reply layers/+lang/ruby/README.org]: I haven't used spacemacs in quite a while. Someone else is welcome to find better bindings, but I doubt I will ever do it.

**Comment by sdwolfz** [comment]
Thank you :heart:! 
And congratulations on your first Spacemacs PR :tada:!

I rebased with latest develop and updated the code in 6302748f993b59eaec8b485c7c920e5c642a02d1. For keybindings I used `SPC m @` since the `@ ` kind of looks like an eye :smiley_cat:. At first this tool did not seem like it would be very usefully, more like a gimmick, but it hit me that this can be used effectively for writing data processing scripts and it's cleaner that using a shell.  

Cherry-picked into develop branch, you can safely delete your branch.

**Pull request closed without merging**